### ISSUE TITLE
GeoJson: style changes would never be updated 

### ISSUE THREAD

**Issue by nguyendh2601** [PaulLeCam/react-leaflet#166]
```
render() {
    const style = _.defaults(location.style, {
      stroke: true,
      color: "#2e5696",
      opacity: 0.6,
      weight: 2,
      dashArray: "1",
      fill: true,
      fillColor: "#2e5696",
      fillOpacity: 0.3,
    });

    return (
      <GeoJson
        style={style}
        data={location.geojson}
        layerContainer={layerContainer}
        map={map}
      >
      </GeoJson>);
  }
}
```

i did init GeoJson as above, the style is correct. But when i change the style later, the GeoJson would not be updated. Took a look at **getPathOptions** in Path.js always return empty style. 
<img width="890" alt="screen shot 2016-05-27 at 1 53 51 pm" src="https://cloud.githubusercontent.com/assets/12975321/15600170/f547a740-241a-11e6-9001-459a84cab360.png">

I'm going with a workaround like this:
```
render() {
    return (
      <GeoJson
        key={location.id}
        style={style}
        data={location.geojson}
        onEachFeature={this.onEachFeature}
        layerContainer={layerContainer}
        map={map}
        {...style}
      >
      </GeoJson>);
}
```

**Comment by nguyendh2601** [comment]
```
L.geoJson( <Object> geojson?, <GeoJSON options> options? )
```
**options** of **GeoJSON**are not **Path options** like **CircleMarker**.
```
L.circleMarker( <LatLng> latlng, <Path options> options? )
```

I think we should change **componentDidUpdate** of ***GeoJson.js*** to
```
componentDidUpdate(prevProps) {
    this.setStyleIfChanged(prevProps.style, this.props.style);
 }
```

What do you think @PaulLeCam @boromisp ? I can create PR for this

**Comment by jgimbel** [comment]
I believe how it currently works is the expected result because of how leafletjs works in the background, even though its not intuitive. Proposing to make the styles object a dynamic prop will break the `resetStyles` function.

For more information...

leafletjs will do something similar to
`this.options = { ...this.options, ...this.options.styles }`
when you initialize. This is because the `options.styles` is considered the source and is never changed. Instead the `options` object is changed when [setStyles](http://leafletjs.com/reference.html#geojson-setstyle) function is used. ([source](https://github.com/Leaflet/Leaflet/blob/5cc9b79c97be3df653474a774f3b06af252a44bc/src/layer/vector/Path.js#L104-110)).

This is because leafletjs holds the initial styles for `resetStyle`. Setting `styles={styles}` will set the **initial** styles, so when `resetStyles` is called, it will use this object to resetStyles. Making the styles prop dynamic will break leafletjs's reference to the initial styles. 

@PaulLeCam A bigger problem tied to GeoJson styles (for me) is that react-leaflet styles are only allowed to be objects, but [leaflet allows styles to be objects or functions](http://leafletjs.com/reference.html#geojson-setstyle) ([source](https://github.com/Leaflet/Leaflet/blob/63f3c851baa282ae31ff8d9522e9e6e0c070c28e/src/layer/GeoJSON.js#L121-125)).

**Comment by PaulLeCam** [comment]
Hi,
Indeed, the `style` prop should be updated as expected if it is a function, that should be fixed in v0.11.6.
